File: openai-agents-python/app/pipeline.py
"""OpenAI Agents SDK feature-delivery pipeline (sequential agents)."""

from __future__ import annotations

import asyncio
import uuid
from pathlib import Path

from agents import (
    Agent,
    AsyncOpenAI,
    OpenAIChatCompletionsModel,
    Runner,
    function_tool,
    set_tracing_disabled,
)

from app.llm_config import resolve_llm_settings
from app.models import RunResult
from app.phase_log import (
    ensure_prerequisites,
    load_artifacts,
    log_phase_start,
    parse_agents,
    validate_agent_selection,
)
from app.tools import Sandbox

ROOT = Path(__file__).resolve().parent.parent
KNOWLEDGE_DIR = ROOT / "knowledge"
OUTPUT_DIR = ROOT / "output"

set_tracing_disabled(True)

SPANISH = (
    "Responde siempre en español. Identificadores de código en inglés cuando sea natural."
)

FRAMEWORK = "openai-agents"


def _build_model(provider: str | None, model: str | None):
    settings = resolve_llm_settings(provider, model)
    client = AsyncOpenAI(api_key=settings.api_key, base_url=settings.base_url)
    model_obj = OpenAIChatCompletionsModel(model=settings.model, openai_client=client)
    return model_obj, settings


def run_feature_delivery(
    request: str,
    *,
    provider: str | None = None,
    model: str | None = None,
    run_id: str | None = None,
    agents: str | list[str] | None = None,
) -> RunResult:
    return asyncio.run(
        _run_async(
            request,
            provider=provider,
            model=model,
            run_id=run_id,
            agents=agents,
        )
    )


async def _run_async(
    request: str,
    *,
    provider: str | None,
    model: str | None,
    run_id: str | None,
    agents: str | list[str] | None,
) -> RunResult:
    selected = parse_agents(agents)
    run_id = run_id or uuid.uuid4().hex[:10]
    output_dir = OUTPUT_DIR / run_id
    sandbox = Sandbox(output_dir, KNOWLEDGE_DIR)
    validate_agent_selection(selected, sandbox)
    model_obj, settings = _build_model(provider, model)

    @function_tool
    def search_knowledge(query: str) -> str:
        """Search local engineering knowledge docs."""
        return sandbox.search_knowledge(query)

    @function_tool
    def web_search(query: str) -> str:
        """Search the web via DuckDuckGo (no API key required)."""
        return sandbox.web_search(query)

    @function_tool
    def write_file(path: str, content: str) -> str:
        """Write a file inside the run sandbox."""
        return sandbox.write_file(path, content)

    @function_tool
    def write_mermaid(path: str, content: str) -> str:
        """Write a Mermaid diagram file under diagrams/."""
        return sandbox.write_mermaid(path, content)

    @function_tool
    def generate_image(prompt: str, path: str) -> str:
        """Generate a PNG mockup via OpenAI Images (requires OPENAI_API_KEY)."""
        return sandbox.generate_image(prompt, path)

    @function_tool
    def read_file(path: str) -> str:
        """Read a file from the run sandbox."""
        return sandbox.read_file(path)

    @function_tool
    def list_files(prefix: str = "") -> str:
        """List files in the run sandbox."""
        return sandbox.list_files(prefix)

    tools = [
        search_knowledge,
        web_search,
        write_file,
        write_mermaid,
        generate_image,
        read_file,
        list_files,
    ]

    agent_map = {
        "researcher": Agent(
            name="Researcher",
            instructions=(
                f"{SPANISH} Eres el Researcher. Usa web_search y search_knowledge. "
                "Escribe research.md con fuentes, resumen e implicaciones."
            ),
            model=model_obj,
            tools=tools,
        ),
        "planner": Agent(
            name="Planner",
            instructions=(
                f"{SPANISH} Eres el Planner. Usa search_knowledge si ayuda. "
                "Escribe plan.md con criterios de aceptación y tareas."
            ),
            model=model_obj,
            tools=tools,
        ),
        "designer": Agent(
            name="Designer",
            instructions=(
                f"{SPANISH} Eres el Designer. Escribe design.md (componentes, APIs, datos). "
                "OBLIGATORIO: diagrama de componentes en fence ```mermaid con flowchart TD/LR. "
                "Etiquetas en texto plano (sin HTML ni <br>; sin comillas dobles). "
                "PROHIBIDO: ASCII/textual, sequenceDiagram, classDiagram."
            ),
            model=model_obj,
            tools=tools,
        ),
        "diagrammer": Agent(
            name="Diagrammer",
            instructions=(
                f"{SPANISH} Eres el Diagrammer. Usa write_mermaid para "
                "diagrams/architecture.mmd y diagrams/sequence.mmd. "
                "REGLA: cada archivo debe empezar con 'flowchart TD' o 'flowchart LR' (o 'graph TD'). "
                "NO uses sequenceDiagram ni classDiagram. "
                "sequence.mmd = flujo temporal con flowchart; architecture.mmd = componentes."
            ),
            model=model_obj,
            tools=tools,
        ),
        "illustrator": Agent(
            name="Illustrator",
            instructions=(
                f"{SPANISH} Eres el Illustrator. Usa generate_image para 1-2 imágenes en assets/. "
                "Si hay Design de producto, prioriza mockups UI; si no, genera la imagen pedida."
            ),
            model=model_obj,
            tools=tools,
        ),
        "coder": Agent(
            name="Coder",
            instructions=(
                f"{SPANISH} Eres el Coder. Implementa bajo src/ con write_file."
            ),
            model=model_obj,
            tools=tools,
        ),
        "reviewer": Agent(
            name="Reviewer",
            instructions=(
                f"{SPANISH} Eres el Reviewer. Lee plan.md, design.md y src/. "
                "Escribe review.md con veredicto approve|request_changes|comment."
            ),
            model=model_obj,
            tools=tools,
        ),
    }

    artifacts = load_artifacts(sandbox)
    plan = str(artifacts["plan"])
    design = str(artifacts["design"])
    research = str(artifacts["research"])

    def _prompts() -> dict[str, str]:
        return {
            "researcher": f"Feature request:\n{request}\n\nEscribe research.md completo.",
            "planner": (
                f"Feature request:\n{request}\n\nResearch:\n{research}\n\nEscribe plan.md completo."
            ),
            "designer": f"Feature request:\n{request}\n\nPlan:\n{plan}\n\nEscribe design.md.",
            "diagrammer": (
                f"Feature request:\n{request}\n\nPlan:\n{plan}\n\nDesign:\n{design}\n\n"
                "Crea diagrams/architecture.mmd y diagrams/sequence.mmd con write_mermaid. "
                "Usa solo flowchart TD/LR o graph TD (nunca sequenceDiagram)."
            ),
            "illustrator": (
                f"Request:\n{request}\n\nResearch:\n{research}\n\nDesign:\n{design}\n\n"
                "Usa generate_image para 1-2 imágenes en assets/ (prompts en inglés). "
                "Si hay Design de producto, prioriza mockups UI; si no, genera la imagen "
                "pedida usando Research/Request."
            ),
            "coder": (
                f"Feature request:\n{request}\n\nPlan:\n{plan}\n\nDesign:\n{design}\n\n"
                "Implementa archivos bajo src/."
            ),
            "reviewer": f"Feature request:\n{request}\n\nRevisa artefactos y escribe review.md.",
        }

    total = len(selected)
    for index, phase in enumerate(selected, start=1):
        ensure_prerequisites(phase, sandbox)
        log_phase_start(phase, index, total, run_id, FRAMEWORK)

        artifacts = load_artifacts(sandbox)
        plan = str(artifacts["plan"])
        design = str(artifacts["design"])
        research = str(artifacts["research"])

        await Runner.run(agent_map[phase], _prompts()[phase], max_turns=40)

    def _read(path: str) -> str:
        text = sandbox.read_file(path)
        return "" if text.startswith("ERROR") else text

    artifacts = load_artifacts(sandbox)
    result = RunResult(
        run_id=run_id,
        output_dir=str(output_dir),
        request=request,
        research=_read("research.md"),
        plan=_read("plan.md"),
        design=_read("design.md"),
        review=_read("review.md"),
        files=sandbox.list_written_files(),
        diagrams=list(artifacts["diagrams"]),  # type: ignore[arg-type]
        assets=list(artifacts["assets"]),  # type: ignore[arg-type]
        provider=settings.provider,
        model=settings.model,
    )
    result.write_summary()
    result.announce()
    return result

File: ai-sdk-typescript/src/pipeline.ts
import "dotenv/config";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { randomBytes } from "node:crypto";
import { createOpenAI } from "@ai-sdk/openai";
import { generateText, tool, stepCountIs } from "ai";
import { z } from "zod";
import { resolveLlmSettings } from "./llmConfig.js";
import { Sandbox } from "./tools.js";
import type { RunResult } from "./models.js";
import { announceResults, writeSummary } from "./models.js";
import {
  ensurePrerequisites,
  loadArtifacts,
  logPhaseStart,
  parseAgents,
  validateAgentSelection,
  type Phase,
} from "./phaseLog.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
export const ROOT = path.resolve(__dirname, "..");
export const KNOWLEDGE_DIR = path.join(ROOT, "knowledge");
export const OUTPUT_DIR = path.join(ROOT, "output");

const SPANISH =
  "Responde siempre en español. Identificadores de código en inglés cuando sea natural.";

const FRAMEWORK = "ai-sdk";

/** Max tool-loop steps per phase (generateText stopWhen). */
const PHASE_MAX_STEPS = 8;

export type RunOptions = {
  provider?: string;
  model?: string;
  runId?: string;
  agents?: string | string[];
};

function buildModel(provider?: string, model?: string) {
  const settings = resolveLlmSettings(provider, model);
  const openai = createOpenAI({
    apiKey: settings.apiKey,
    baseURL: settings.baseUrl,
  });
  return { model: openai(settings.model), settings };
}

function buildTools(sandbox: Sandbox) {
  return {
    search_knowledge: tool({
      description: "Search local engineering knowledge docs",
      inputSchema: z.object({ query: z.string() }),
      execute: async ({ query }) => sandbox.searchKnowledge(query),
    }),
    web_search: tool({
      description: "Search the web via DuckDuckGo (no API key required)",
      inputSchema: z.object({ query: z.string() }),
      execute: async ({ query }) => sandbox.webSearch(query),
    }),
    write_file: tool({
      description: "Write a file inside the run sandbox",
      inputSchema: z.object({ path: z.string(), content: z.string() }),
      execute: async ({ path: filePath, content }) =>
        sandbox.writeFile(filePath, content),
    }),
    write_mermaid: tool({
      description: "Write a Mermaid diagram file under diagrams/",
      inputSchema: z.object({ path: z.string(), content: z.string() }),
      execute: async ({ path: filePath, content }) =>
        sandbox.writeMermaid(filePath, content),
    }),
    generate_image: tool({
      description:
        "Generate a PNG mockup via OpenAI Images (requires OPENAI_API_KEY)",
      inputSchema: z.object({ prompt: z.string(), path: z.string() }),
      execute: async ({ prompt, path: filePath }) =>
        sandbox.generateImage(prompt, filePath),
    }),
    read_file: tool({
      description: "Read a file from the run sandbox",
      inputSchema: z.object({ path: z.string() }),
      execute: async ({ path: filePath }) => sandbox.readFile(filePath),
    }),
    list_files: tool({
      description: "List files in the run sandbox",
      inputSchema: z.object({ prefix: z.string().optional() }),
      execute: async ({ prefix }) => sandbox.listFiles(prefix ?? ""),
    }),
  };
}

type PhaseContext = {
  request: string;
  research: string;
  plan: string;
  design: string;
};

type PhaseDef = {
  system: string;
  prompt: (ctx: PhaseContext) => string;
};

const PHASE_DEFS: Record<Phase, PhaseDef> = {
  researcher: {
    system: `${SPANISH} Eres el Researcher. Usa web_search y search_knowledge. Escribe research.md con write_file.`,
    prompt: (ctx) =>
      `Feature request:\n${ctx.request}\nEscribe research.md completo.`,
  },
  planner: {
    system: `${SPANISH} Eres el Planner. Usa search_knowledge si ayuda. Escribe plan.md con write_file.`,
    prompt: (ctx) =>
      `Feature request:\n${ctx.request}\n\nResearch:\n${ctx.research}\nEscribe plan.md completo.`,
  },
  designer: {
    system: `${SPANISH} Eres el Designer. Escribe design.md con write_file (componentes, APIs, datos). OBLIGATORIO: diagrama en fence \`\`\`mermaid con flowchart TD/LR. Etiquetas en texto plano (sin HTML ni <br>). PROHIBIDO: ASCII/textual, sequenceDiagram, classDiagram.`,
    prompt: (ctx) =>
      `Feature:\n${ctx.request}\n\nPlan:\n${ctx.plan}\nEscribe design.md con diagrama Mermaid de componentes (no ASCII).`,
  },
  diagrammer: {
    system: `${SPANISH} Eres el Diagrammer. Usa write_mermaid para diagrams/architecture.mmd y diagrams/sequence.mmd. REGLA: cada archivo debe empezar con 'flowchart TD' o 'flowchart LR' (o 'graph TD'). NO uses sequenceDiagram ni classDiagram. sequence.mmd = flujo temporal con flowchart; architecture.mmd = componentes.`,
    prompt: (ctx) =>
      `Feature:\n${ctx.request}\n\nPlan:\n${ctx.plan}\n\nDesign:\n${ctx.design}\n` +
      `Crea diagrams/architecture.mmd y diagrams/sequence.mmd con write_mermaid. ` +
      `Usa solo flowchart TD/LR o graph TD (nunca sequenceDiagram).`,
  },
  illustrator: {
    system: `${SPANISH} Eres el Illustrator. Usa generate_image para 1-2 imágenes en assets/. Si hay Design de producto, prioriza mockups UI; si no, genera la imagen pedida.`,
    prompt: (ctx) =>
      `Request:\n${ctx.request}\n\nResearch:\n${ctx.research}\n\nDesign:\n${ctx.design}\n` +
      `Usa generate_image para 1-2 imágenes en assets/ (prompts en inglés). ` +
      `Si hay Design de producto, prioriza mockups UI; si no, genera la imagen ` +
      `pedida usando Research/Request.`,
  },
  coder: {
    system: `${SPANISH} Eres el Coder. Implementa bajo src/ con write_file.`,
    prompt: (ctx) =>
      `Feature:\n${ctx.request}\n\nPlan:\n${ctx.plan}\n\nDesign:\n${ctx.design}\n` +
      `Implementa archivos bajo src/ con write_file.`,
  },
  reviewer: {
    system: `${SPANISH} Eres el Reviewer. Lee artefactos con tools y escribe review.md con veredicto.`,
    prompt: (ctx) =>
      `Feature:\n${ctx.request}\nLee plan.md, design.md y src/. Escribe review.md ` +
      `con hallazgos y veredicto approve|request_changes|comment. Usa search_knowledge para checklist.`,
  },
};

function buildPhaseRunners(
  sandbox: Sandbox,
  model: ReturnType<ReturnType<typeof createOpenAI>>,
  tools: ReturnType<typeof buildTools>,
) {
  async function runPhase(phase: Phase, ctx: PhaseContext) {
    const def = PHASE_DEFS[phase];
    const result = await generateText({
      model,
      system: def.system,
      prompt: def.prompt(ctx),
      stopWhen: stepCountIs(PHASE_MAX_STEPS),
      tools,
    });
    return result.text ?? "";
  }

  const runners: Record<
    Phase,
    (ctx: PhaseContext) => Promise<{
      research: string;
      plan: string;
      design: string;
      review: string;
    }>
  > = {
    researcher: async (ctx) => {
      const text = await runPhase("researcher", ctx);
      const researchFile = sandbox.readFile("research.md");
      const research = researchFile.startsWith("ERROR") ? text : researchFile;
      return { research, plan: ctx.plan, design: ctx.design, review: "" };
    },
    planner: async (ctx) => {
      const text = await runPhase("planner", ctx);
      const planFile = sandbox.readFile("plan.md");
      const plan = planFile.startsWith("ERROR") ? text : planFile;
      return { research: ctx.research, plan, design: ctx.design, review: "" };
    },
    designer: async (ctx) => {
      const text = await runPhase("designer", ctx);
      const designFile = sandbox.readFile("design.md");
      const design = designFile.startsWith("ERROR") ? text : designFile;
      return { research: ctx.research, plan: ctx.plan, design, review: "" };
    },
    diagrammer: async (ctx) => {
      await runPhase("diagrammer", ctx);
      return {
        research: ctx.research,
        plan: ctx.plan,
        design: ctx.design,
        review: "",
      };
    },
    illustrator: async (ctx) => {
      await runPhase("illustrator", ctx);
      return {
        research: ctx.research,
        plan: ctx.plan,
        design: ctx.design,
        review: "",
      };
    },
    coder: async (ctx) => {
      await runPhase("coder", ctx);
      return {
        research: ctx.research,
        plan: ctx.plan,
        design: ctx.design,
        review: "",
      };
    },
    reviewer: async (ctx) => {
      const text = await runPhase("reviewer", ctx);
      const reviewFile = sandbox.readFile("review.md");
      const review = reviewFile.startsWith("ERROR") ? text : reviewFile;
      return {
        research: ctx.research,
        plan: ctx.plan,
        design: ctx.design,
        review,
      };
    },
  };

  return runners;
}

export async function runFeatureDelivery(
  request: string,
  options: RunOptions = {},
): Promise<RunResult> {
  const selected = parseAgents(options.agents);
  const { model, settings } = buildModel(options.provider, options.model);
  const runId = options.runId ?? randomBytes(5).toString("hex");
  const outputDir = path.join(OUTPUT_DIR, runId);
  const sandbox = new Sandbox(outputDir, KNOWLEDGE_DIR);
  validateAgentSelection(selected, sandbox);
  const tools = buildTools(sandbox);
  const runners = buildPhaseRunners(sandbox, model, tools);

  const artifacts = loadArtifacts(sandbox);
  let ctx: PhaseContext = {
    request,
    research: artifacts.research,
    plan: artifacts.plan,
    design: artifacts.design,
  };
  let review = artifacts.review;

  const total = selected.length;
  for (let index = 0; index < selected.length; index++) {
    const phase = selected[index];
    ensurePrerequisites(phase, sandbox);
    logPhaseStart(phase, index + 1, total, runId, FRAMEWORK);
    const result = await runners[phase](ctx);
    ctx = {
      request,
      research: result.research,
      plan: result.plan,
      design: result.design,
    };
    if (result.review) review = result.review;
  }

  const clean = (v: string) => (v?.startsWith("ERROR") ? "" : v || "");
  const finalArtifacts = loadArtifacts(sandbox);

  const runResult: RunResult = {
    run_id: runId,
    output_dir: outputDir,
    request,
    research: clean(ctx.research || sandbox.readFile("research.md")),
    plan: clean(ctx.plan || sandbox.readFile("plan.md")),
    design: clean(ctx.design || sandbox.readFile("design.md")),
    review: clean(review || sandbox.readFile("review.md")),
    files: sandbox.listWrittenFiles(),
    diagrams: finalArtifacts.diagrams,
    assets: finalArtifacts.assets,
    provider: settings.provider,
    model: settings.model,
  };
  writeSummary(runResult);
  announceResults(runResult);
  return runResult;
}
